### PR TITLE
[mlir][test] Add TD example for peel+vectorize (depthwise conv)

### PR DESCRIPTION
**Pull request by banach-space**
Adds an example that combines loop peeling and scalable vectorisation of
`linalg.depthwise_conv_2d_nhwc_hwc`. This is similar to
transform-op-peel-and-vectorize.mlir and is meant to demonstrate how to
avoid masking when vectorising using scalable vectors.

### CODE REVIEW

**Comment by banach-space** [comment]
CC @nujaa

**Comment by nujaa** [comment]
Hi, thanks for the example, very insightful. Nothing to complain really, just wondering if the canonicalization would work in dynamic sizes. 
In opposition to the matmul usecase, you run a canonicalization after decompose. Is it required by decompose ? ( I am not too familiar with the op)

File: mlir/test/Dialect/Linalg/transform-op-peel-and-vectorize-conv.mlir [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    %tiled_linalg_op_1, %loops_2:2 = transform.structured.tile_using_for %2[0, 0, 0, 0, 1, 1] : (!transform.any_op) -> (!transform.any_op, !transform.any_op, !transform.any_op)
+
+    // 3. Decompose 2D conv into 2 x 1D conv
+    %3 = transform.structured.match ops{["linalg.depthwise_conv_2d_nhwc_hwc"]} in %loops_1#3 : (!transform.op<"scf.for">) -> !transform.any_op
+    %4 = transform.structured.decompose %3 : (!transform.any_op) -> !transform.any_op
+    %f00 = transform.structured.match ops{["func.func"]} in %root
+        : (!transform.any_op) -> !transform.any_op
+    transform.apply_patterns to %f00 {
+        transform.apply_patterns.canonicalization
+        transform.apply_patterns.linalg.tiling_canonicalization
```
[the remark this answers is not in the document]
Comment (banach-space) [reply mlir/test/Dialect/Linalg/transform-op-peel-and-vectorize-conv.mlir:77]: > Isn't tiling_canonicalization a subset of canonicalization ?

Most likely :)

https://mlir.llvm.org/docs/Dialects/Transform/#transformapply_patternslinalgtiling_canonicalization-transformapplytilingcanonicalizationpatternsop

> Isn't canonicalization enough

Yup, these patterns are not needed here. Thanks for pointing out!

**Comment by banach-space** [comment]
>  wondering if the canonicalization would work in dynamic sizes. 

It should, but I won't have the bandwidth to check. It's not hard to update examples like this, just a bit fiddly, and I need to prioritise sth else in the coming weeks 😅 

> Is it required by decompose ? ( I am not too familiar with the op)

Yes. The vectoriser can only vectorise 1D convs. We deal with 2D convs by decomposing them into 1D convs. That's like a "vectoriser pre-processing" stage.

**Review by nujaa — APPROVED** [review]
LGTM.

**Pull request merged**